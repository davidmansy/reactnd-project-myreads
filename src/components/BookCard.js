import React, { Component } from 'react';
import ShelMoveMenu from './ShelfMoveMenu';

class BookCard extends Component {
  render() {
    const { book } = this.props;
    const bookCoverStyle = {
      width: 128,
      height: 193,
      backgroundImage: `url(${book.imageLinks.thumbnail})`
    };
    const authors = book.authors.join(' ');

    return (
      <div className="book">
        <div className="book-top">
          <div className="book-cover" style={bookCoverStyle} />
          <div className="book-shelf-changer">
            <ShelMoveMenu />
          </div>
        </div>
        <div className="book-title">{book.title}</div>
        {/*TODO: Use a list for the authors?*/}
        <div className="book-authors">{authors}</div>
      </div>
    );
  }
}

export default BookCard;
