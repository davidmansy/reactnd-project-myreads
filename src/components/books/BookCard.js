import React, { Component } from 'react';
import ShelMoveMenu from '../common/ShelfMoveMenu';

const BOOK_DETAILS_HIDE_CLASS = 'book-details-hidden';
const DATA_TRANSFER_ID = 'id';

class BookCard extends Component {
  state = {
    hideBook: false,
    active: false
  };

  onDragStart = (e, id) => {
    e.dataTransfer.setData(DATA_TRANSFER_ID, id);
    setTimeout(() => {
      this.setState({
        hideBook: true
      });
    });
  };

  onDragEnd = e => {
    this.setState({
      hideBook: false
    });
  };

  onMouseEnter = e => {
    if (this.props.draggable === 'true') {
      this.setState({
        active: true
      });
    }
  };

  onMouseLeave = e => {
    if (this.props.draggable === 'true') {
      this.setState({
        active: false
      });
    }
  };

  render() {
    const { book, handleShelfChange, draggable = 'false' } = this.props;
    const thumbnail = book.imageLinks ? book.imageLinks.thumbnail : null;
    const bookCoverStyle = {
      width: 128,
      height: 193,
      backgroundImage: `url(${thumbnail})`
    };
    const containerClass = this.state.hideBook ? BOOK_DETAILS_HIDE_CLASS : '';
    const bookClass = this.state.active ? 'book-active' : '';

    return (
      <div
        className={`book ${bookClass}`}
        onDragStart={e => {
          this.onDragStart(e, book.id);
        }}
        onDragEnd={this.onDragEnd}
        draggable={draggable}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
      >
        <div className={`book-details-container ${containerClass}`}>
          <div className="book-top">
            <div className="book-cover" style={bookCoverStyle} />
            <div className="book-shelf-changer">
              <ShelMoveMenu
                selectedShelf={book.shelf}
                handleChange={event => {
                  handleShelfChange(book, event.target.value);
                }}
              />
            </div>
          </div>
          <div className="book-title">{book.title}</div>
          <div className="book-authors">{book.authors}</div>
        </div>
      </div>
    );
  }
}

export default BookCard;
