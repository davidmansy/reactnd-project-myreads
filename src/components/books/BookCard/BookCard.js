import React, { Component } from 'react';
import ShelMoveMenu from '../../common/ShelfMoveMenu';
import PropTypes from 'prop-types';

const BOOK_CLASS = 'book';
const BOOK_ACTIVE_CLASS = 'book-active';
const BOOK_DETAILS_CONTAINER_CLASS = 'book-details-container';
const BOOK_DETAILS_HIDDEN_CONTAINER_CLASS = 'book-details-container-hidden';
const DATA_TRANSFER_ID = 'id';

class BookCard extends Component {
  state = {
    hidden: false,
    active: false
  };

  onDragStart = (e, id) => {
    e.dataTransfer.setData(DATA_TRANSFER_ID, id);
    setTimeout(() => {
      this.setState({
        hidden: true
      });
    });
  };

  onDragEnd = e => {
    this.setState({
      hidden: false
    });
  };

  onMouseEnter = e => {
    if (this.props.draggable === 'true') {
      this.setState({
        active: true
      });
    }
  };

  onMouseLeave = e => {
    if (this.props.draggable === 'true') {
      this.setState({
        active: false
      });
    }
  };

  render() {
    const { book, handleShelfChange, draggable } = this.props;
    const thumbnail = book.imageLinks ? book.imageLinks.thumbnail : null;
    const bookCoverStyle = {
      width: 128,
      height: 193,
      backgroundImage: `url(${thumbnail})`
    };
    const containerClass = this.state.hidden
      ? `${BOOK_DETAILS_CONTAINER_CLASS} ${BOOK_DETAILS_HIDDEN_CONTAINER_CLASS}`
      : `${BOOK_DETAILS_CONTAINER_CLASS}`;
    const bookClass = this.state.active
      ? `${BOOK_CLASS} ${BOOK_ACTIVE_CLASS}`
      : `${BOOK_CLASS}`;

    return (
      <div
        className={bookClass}
        onDragStart={e => {
          this.onDragStart(e, book.id);
        }}
        onDragEnd={this.onDragEnd}
        draggable={draggable}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
      >
        <div className={containerClass}>
          <div className="book-top">
            <div className="book-cover" style={bookCoverStyle} />
            <div className="book-shelf-changer">
              <ShelMoveMenu
                selectedShelf={book.shelf}
                handleChange={event => {
                  handleShelfChange(book, event.target.value);
                }}
              />
            </div>
          </div>
          <div className="book-title">{book.title}</div>
          <div className="book-authors">{book.authors}</div>
        </div>
      </div>
    );
  }
}

BookCard.propTypes = {
  book: PropTypes.object.isRequired,
  handleShelfChange: PropTypes.func.isRequired,
  draggable: PropTypes.string
};

BookCard.defaultProps = {
  draggable: 'false'
};

export default BookCard;
