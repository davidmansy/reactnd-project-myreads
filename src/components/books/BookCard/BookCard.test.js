import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import BookCard from './BookCard';
import { MemoryRouter } from 'react-router-dom';
import { cleanup } from 'react-testing-library';

afterEach(cleanup);

describe('BookCard', () => {
  const props = {
    book: { shelf: 'read' },
    handleShelfChange: () => {},
    draggable: 'true'
  };

  it('renders', () => {
    const container = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <BookCard {...props} />
      </MemoryRouter>,
      container
    );
  });

  it('snapshots', () => {
    const testRenderer = TestRenderer.create(
      <MemoryRouter>
        <BookCard {...props} />
      </MemoryRouter>
    );
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
