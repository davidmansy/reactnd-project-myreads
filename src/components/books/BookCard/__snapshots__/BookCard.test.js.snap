// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard snapshots 1`] = `
<div
  className="book"
  draggable="true"
  onDragEnd={[Function]}
  onDragStart={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="book-details-container"
  >
    <div
      className="book-top"
    >
      <div
        className="book-cover"
        style={
          Object {
            "backgroundImage": "url(null)",
            "height": 193,
            "width": 128,
          }
        }
      />
      <div
        className="book-shelf-changer"
      >
        <select
          onChange={[Function]}
          value="read"
        >
          <option
            disabled={true}
            value="move"
          >
            Move to...
          </option>
          <option
            value="wantToRead"
          >
            Want to Read
          </option>
          <option
            value="currentlyReading"
          >
            Currently Reading
          </option>
          <option
            value="read"
          >
            Read
          </option>
          <option
            value="none"
          >
            None
          </option>
        </select>
      </div>
    </div>
    <div
      className="book-title"
    />
    <div
      className="book-authors"
    />
  </div>
</div>
`;
